junebug/plugins/nginx/plugin.py

```diff
@@ -0,0 +1,125 @@
+from urlparse import urljoin
+
+from confmodel import Config
+from confmodel.fields import ConfigText
+
+from junebug.plugin import JunebugPlugin
+from junebug.utils import channel_public_http_properties
+
+
+log = logging.getLogger(__name__)
+DIRNAME = path.dirname(__file__)
+
+
+class NginxPluginConfig(Config):
+    '''Config for :class:`NginxJunebugPlugin`'''
+
+    vhost_file = ConfigText(
+        "The file to write the junebug nginx vhost file to",
+        default='/etc/nginx/sites-enabled/junebug.conf', static=True)
+
+    locations_dir = ConfigText(
+        "The directory to write location block config files to",
+        default='/etc/nginx/includes/junebug/', static=True)
+
+    server_name = ConfigText(
+        "Server name to use for nginx vhost",
+        required=True, static=True)
+
+    vhost_template = ConfigText(
+        "Path to the template file to use for the vhost config",
+        default='%s/vhost.template' % (DIRNAME,), static=True)
+
+    location_template = ConfigText(
+        "Path to the template file to use for each channel's location config",
+        default='%s/location.template' % (DIRNAME,), static=True)
+
+
+class NginxPlugin(JunebugPlugin):
+    '''
+    Manages an nginx virtual host that proxies to the Junebug instance's
+    http-based channels.
+    '''
+
+    def start_plugin(self, config, junebug_config):
+        self.configured_channels = set()
+        self.config = NginxPluginConfig(config)
+        self.vhost_template = read(self.config.vhost_template)
+        self.location_template = read(self.config.location_template)
+        write(self.config.vhost_file, self.get_vhost_config())
+
+    def stop_plugin(self):
+        ensure_removed(self.config.vhost_file)
+
+    def channel_started(self, channel):
+        properties = channel_public_http_properties(channel._properties)
+
+        if properties is not None and properties['enabled']:
+            mkpath(self.config.locations_dir)
+
+            write(
+                self.get_location_path(channel.id),
+                self.get_location_config(properties))
+
+            reload_nginx()
+
+            self.configured_channels.add(channel.id)
+
+    def channel_stopped(self, channel):
+        if channel.id in self.configured_channels:
+            ensure_removed(self.get_location_path(channel.id))
+            self.configured_channels.remove(channel.id)
+
+    def get_vhost_config(self):
+        return self.vhost_template % self.get_vhost_context()
+
+    def get_vhost_context(self):
+        return {
+            'server_name': self.config.server_name,
+            'includes': path.join(self.config.locations_dir, '*.conf')
+        }
+
+    def get_location_config(self, properties):
+        return self.location_template % self.get_location_context(properties)
+
+    def get_location_context(self, properties):
+        web_path = properties['web_path']
+        base_url = 'http://localhost:%s' % (properties['web_port'],)
```

Comment:
Hmm, good point. Agreed that it isn't in this PR's scope, but we should discuss this.